rpc/src/rpc/native/state.rs

```diff
@@ -0,0 +1,195 @@
+                address_storage_access.extend(read_keys);
+            }
+
+            if let Some(written_keys) = trace.storage_written.as_ref() {
+                address_storage_access.extend(written_keys.keys());
+            }
+        }
+    }
+
+    Ok(state_access)
+}
+
+/// Generates the state witness for the given block.
+async fn generate_state_witness(
+    prev_state_root: H256,
+    accounts_state: HashMap<H160, HashSet<H256>>,
+    provider: Arc<Provider<RetryClient<Http>>>,
+    block_number: ethereum_types::U64,
+) -> Result<(
+    PartialTrieBuilder<HashedPartialTrie>,
+    HashMap<H256, PartialTrieBuilder<HashedPartialTrie>>,
+)> {
+    let mut state = PartialTrieBuilder::new(prev_state_root, Default::default());
+    let mut storage_proofs =
+        HashMap::<HashedStorageAddr, PartialTrieBuilder<HashedPartialTrie>>::new();
+
+    let (account_proofs, next_account_proofs) =
+        fetch_proof_data(accounts_state, provider, block_number).await?;
+
+    // Insert account proofs
+    for (address, proof) in account_proofs.into_iter() {
+        state.insert_proof(proof.account_proof.map_into_bytes());
+
+        let storage_mpt =
+            storage_proofs
+                .entry(keccak256(address).into())
+                .or_insert(PartialTrieBuilder::new(
+                    proof.storage_hash,
+                    Default::default(),
+                ));
+        for proof in proof.storage_proof {
+            storage_mpt.insert_proof(proof.proof.map_into_bytes());
+        }
+    }
+
+    // Insert short node variants from next proofs
+    for (address, proof) in next_account_proofs.into_iter() {
+        state.insert_short_node_variants_from_proof(proof.account_proof.map_into_bytes());
+
+        if let Some(storage_mpt) = storage_proofs.get_mut(&keccak256(address).into()) {
+            for proof in proof.storage_proof {
+                storage_mpt.insert_short_node_variants_from_proof(proof.proof.map_into_bytes());
+            }
+        }
+    }
+
+    Ok((state, storage_proofs))
+}
+
+/// Fetches the proof data for the given accounts and associated storage keys.
+async fn fetch_proof_data(
+    accounts_state: HashMap<H160, HashSet<H256>>,
+    provider: Arc<Provider<RetryClient<Http>>>,
+    block_number: ethereum_types::U64,
+) -> Result<
+    (
+        Vec<(H160, EIP1186ProofResponse)>,
+        Vec<(H160, EIP1186ProofResponse)>,
+    ),
+    anyhow::Error,
+> {
+    let account_proofs_fut = accounts_state
+        .clone()
+        .into_iter()
+        .map(|(address, keys)| {
+            let provider = provider.clone();
+            async move {
+                let proof = provider
+                    .get_proof(
+                        address,
+                        keys.into_iter().collect(),
+                        Some((block_number - 1).into()),
+                    )
+                    .map_err(|e| anyhow!("Failed to get proof for account: {:?}", e))
```

Comment:
```rust
use anyhow::Context as __;
```
```suggestion
                    .context("Failed to get proof for account")
```

Anyhow will preserve the source error in it's cause chain. This is the whole point of using anyhow